src/hupper/reloader.py

```diff
@@ -133,21 +133,27 @@ def run(self):
         """
         Execute the reloader forever, blocking the current thread.
 
-        This will invoke ``sys.exit(1)`` if interrupted.
+        This will invoke ``sys.exit(1)`` if interrupted and
+        the process didn't handle the interrupt gracefully.
 
         """
+        exitcode = 1
         with self._setup_runtime():
             while True:
-                result = self._run_worker()
-                start = time.time()
-                if result == WorkerResult.WAIT:
-                    result = self._wait_for_changes()
+                result, worker_exitcode = self._run_worker()
                 if result == WorkerResult.EXIT:
+                    if worker_exitcode is not None:
+                        exitcode = worker_exitcode
```

Comment:
I don't think it's possible for the worker to return a `None` exitcode so this check should be unnecessary and it can just be `result, exitcode = self._run_worker()`.

Reply:
This was a bit defensive because I wasn't 100% certain this being guaranteed to be non-None, and I didn't want to swallow an error by exiting with None (which will exit with 0).

Reply:
FWIW the return type of `ipc.wait` is `int | None`

Reply:
Yep the return code is `None` until the process exits so if you pass in a`timeout=0` it uses a`poll` which might do that. The `.join()` in `_run_worker` is doing a  blocking wait before returning and it’ll never be `None`.